kernel-plugins/test/foundry/BLSValidator.test.sol

```diff
@@ -0,0 +1,181 @@
+        beneficiary = payable(address(makeAddr("beneficiary")));
+    }
+
+    function test_mode_2_bls() external {
+        BLSValidator blsValidator = new BLSValidator();
+        // Note: With the Kernel wallet validation and execution are seperate
+        // Executors are plugins that add custom functions to Kernel and each function
+        // is tied to a validator.  In this test we will be enabling the blsValidator and
+        // tying the TextExecutor to it.
+        TestExecutor testExecutor = new TestExecutor();
+        UserOperation memory op =
+            entryPoint.fillUserOp(address(kernel), abi.encodeWithSelector(TestExecutor.doNothing.selector));
+
+        // BLS public/private key pair and signatures created using the '@thehubbleproject/bls' library.
+        // Hard coded in the test because I don't know of a way to do this in solidity.
+        uint256[4] memory publicKey = [
+            0x004b1d8408dcc643647e4f32a761853e873cd1da8ffc40f03b00647484b3498a,
+            0x248b9979254108c9cbb2005739dc693f1694b7b2058942114a0ab4aa81723a6d,
+            0x0edd89947147f52246e1dc3092b62c9020afca38d470b3b827e267c66350da93,
+            0x2d3f8d5cc58f02219b73b70a0acf33778335e7b5f5c7bf98a1d596b9270e92bb
+        ];
+
+        bytes memory enableData = abi.encodePacked(
+            publicKey,
+            uint48(0),
+            TestExecutor.doNothing.selector,
+            uint48(0),
+            uint48(0),
+            uint32(16)
+        );
+
+        bytes32 digest = getTypedDataHash(
+            address(kernel),
+            TestExecutor.doNothing.selector,
+            0,
+            0,
+            address(blsValidator),
+            address(testExecutor),
+            enableData
+        );
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(ownerKey, digest);
+
+        op.signature = abi.encodePacked(
+            bytes4(0x00000002),
+            uint48(0),
+            uint48(0),
+            address(blsValidator),
+            address(testExecutor),
+            uint256(enableData.length),
+            enableData,
+            uint256(65),
+            r,
+            s,
+            v
+        );
+
+        // The signature is hardcoded because the userOpHash is signed outside
+        // the test is a seperate script.
+        // bytes32 hash = entryPoint.getUserOpHash(op);
+        uint256[2] memory signature = [
+            0x238bd40a43095cc868aa13820ced70e0deedc62f588a029cd9aad8a6b72a2283,
+            0x21f8022a28193d60e8f28bb4ce35086c28b1bbe5bb5bf1564bc7b63a2e544f54
+        ];
+
+        // First signature is the ecdsa signature for the default (ecdsa) validator. We
+        // need that signature to validate the BLSValidator for the kernel account.
+        // Second signature is the BLS signature for the BLSValidator. We are calling
+        // the TestExecutor with the BLSValidator as the validator.
+        op.signature = bytes.concat(op.signature, abi.encodePacked(signature));
```

Comment:
Can you explain why we need to create `op.signature` and then concat it with the hard-coded bls sig? Not sure I'm quite understanding the kernel flow

Reply:
Yeah, I copied another test for this, but I'll try my best to explain it.  When you enable a validation plugin for a kernel wallet, you associate that validation with an execution function.  After doing so you can call that execution function from your kernel wallet with the associated validation method.

In this test I'm doing the following:
- Create a UserOp to call the `TestExecutor.doNothing` function from my existing kernel wallet
- Pass the `blsPublicKey` and the `TestExecutor.doNothing.selector` in the enable data.  This will associate the blsPublicKey with this function.
- enabledata, validator, and executor are signed with a ECDSA signature and set to the `op.signature` value.

I think at this point we have basically created an ecdsa signature that is enabling the new validator.  

- Then we concat the ecdsa sig and a BLS signature of the userOp hash.
- This BLS signature will be used to call `TestExecutor.doNothing` after the BLSValidator is enabled with the ecdsa signature.

Reply:
I need to see if I can find some documentation on this structure.  I basically went off of how other tests were set up and just figured it out.  I can also create more tests to test out other situations.  For example, can we enable a validator without calling the execution function?  Also, I want a test of calling to execution function multiple times after enabling the validator.

Reply:
Ok interesting. I think that flow makes sense as you've explained it.

The ECDSA sig part feels quite weird to me. Few follow up questions to clarify:
1. Is the ECDSA signer equivalent to the owner of the kernel wallet?
2. Could you in theory swap out the ECDSA sig for another sig? Like a webauthn sig, or another bls sig?
3. Does the ECDSA signature need to be done only to enable/disable the validator? Or is it always needed to send a user op using the validator too?
4. Is the execution function a singleton too? Or is that specific to an account?

Reply:
Good questions - I'll try to answer them:

1. Yeah.  The default validator for the new Kernel wallet in my tests is an ECDSA validator.  So whoever owns that private key owns that Kernel wallet.
2. Yeah, I'm only using the ECDSA sig because it is the default validator when I created the wallet.  I use an ECDSAKernelFactory to generate the wallet in the test.  Theoretically, I could create a WebAuthNKernelFactory that creates the kernel with a webauthn validator as the default validator.

Some good notes from the documentation on default validators
> Each Kernel account must have one "default validator." A default validator is special in a few ways:
> - It's the validator used in "sudo mode" (see overview). In other words, when other validators are not used, the default validator is used by default (duh).
> - It's the validator used in "enable mode" to check if it approves of the enabling of another validator.
> - It's the validator used for checking signatures for ERC-1271.
> The second and third points are the reasons why a default validator must implement a validateSignature method.

3. It is just needed to enable/disable the validator (I should also test out disabling a validator).  Once the validator is set for an executor, you just need the applicable signature.
4. I think executors can be anything you want. You can use an existing executor or deploy your own.  It is just like adding a custom function to your wallet.  The only requirement is that executors don't touch any storage used by the Kernel wallet.  Here's an example they created for erc721 functions https://github.com/zerodevapp/kernel/blob/main/src/executor/TokenActions.sol

Reply:
ok all makes sense, thanks for the detailed answers!